### PR TITLE
feat: add webauthn 2fa and login

### PR DESCRIPTION
**Pull request by asbiin**
image

After a first normal login, you can use passwordless login:
image

Registration
image
image

### CODE REVIEW

File: app/Actions/Fortify/RedirectIfTwoFactorAuthenticatable.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * The login rate limiter instance.
+     *
+     * @var \Laravel\Fortify\LoginRateLimiter
+     */
+    protected $limiter;
+
+    /**
+     * Create a new controller instance.
```
Comment (djaiss) [inline app/Actions/Fortify/RedirectIfTwoFactorAuthenticatable.php:30]: Wrong comment, i think.

File: app/Actions/AttemptToAuthenticateWebauthn.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    /**
+     * The login rate limiter instance.
+     *
+     * @var \LaravelWebauthn\Services\LoginRateLimiter
+     */
+    protected LoginRateLimiter $limiter;
+
+    /**
+     * Create a new controller instance.
```
Comment (djaiss) [inline app/Actions/AttemptToAuthenticateWebauthn.php:31]: Wrong comment, i think.

File: app/Actions/Fortify/TwoFactorChallengeView.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        $userId = $request->session()->get('login.id');
+        $user = User::find($userId);
+
+        $data = [];
+        if ($user !== null) {
+            $data['publicKey'] = Webauthn::prepareAssertion($user);
+        }
+
+        return Inertia::render('Auth/TwoFactorChallenge', $data + [
+            'two_factor' => optional($user)->two_factor_secret && ! is_null(optional($user)->two_factor_confirmed_at),
```
Comment (djaiss) [inline app/Actions/Fortify/TwoFactorChallengeView.php:29]: This line is super hard to read. It's optional + not is null... We should have a variable here, and the variable defined above, so it's easier to understand.

File: app/Actions/Jetstream/UserProfile.php [hunk 4]
```diff
@@ -21,9 +21,22 @@ public function __invoke(Request $request, array $data): array
             return [$provider => config("services.$provider.name") ?? __("auth.login_provider_{$provider}")];
         });
 
+        $webauthnKeys = $request->user()->webauthnKeys()
```
Comment (djaiss) [inline app/Actions/Jetstream/UserProfile.php:24]:
```suggestion
        $webAuthnKeys = $request->user()->webauthnKeys()
```

File: app/Actions/Jetstream/UserProfile.php [hunk 5]
```diff
@@ -21,9 +21,22 @@ public function __invoke(Request $request, array $data): array
...
+                    'type' => $key->type,
+                    'last_active' => $key->updated_at->diffForHumans(),
+                ];
+            })
+            ->toArray();
+
         $data['providers'] = $providers;
         $data['providersName'] = $providersName;
         $data['userTokens'] = $request->user()->usertokens()->get();
+        $data['webauthnKeys'] = $webauthnKeys;
```
Comment (djaiss) [inline app/Actions/Jetstream/UserProfile.php:39]:
```suggestion
        $data['webauthnKeys'] = $webAuthnKeys;
```

File: app/Http/Controllers/Auth/LoginController.php [hunk 6]
```diff
@@ -24,7 +27,18 @@ public function __invoke(Request $request): Response
             return [$provider => config("services.$provider.name") ?? __("auth.login_provider_{$provider}")];
         });
 
-        return Inertia::render('Auth/Login', [
+        $webauthnRemember = $request->cookie('webauthn_remember');
+        $data = [];
+        if ($webauthnRemember) {
+            if (($user = User::find($webauthnRemember)) && $user->webauthnKeys()->count() > 0) {
```
Comment (djaiss) [inline app/Http/Controllers/Auth/LoginController.php:33]: Are you sure that works?
You instantiate the $user variable and right after, you query it with `$user->something`? If the user can't be found, the latter won't work.

File: app/Actions/Fortify/TwoFactorChallengeView.php (hunk 3 above)
Comment (asbiin) [reply app/Actions/Fortify/TwoFactorChallengeView.php:29]: Actually it's a copy/paste code from laravel/fortify, so I wouldn't bother too much

File: app/Actions/Jetstream/UserProfile.php (hunk 4 above)
Comment (asbiin) [reply app/Actions/Jetstream/UserProfile.php:24]: No, it's Webauthn

File: app/Http/Controllers/Auth/LoginController.php (hunk 6 above)
Comment (asbiin) [reply app/Http/Controllers/Auth/LoginController.php:33]: Yes it is working, the first parenthesis `($user = ...)` is evaluated before, and if the user is not found, then the `if` result is `false`.

**Pull request merged**